viomem/sys/Device.c

```diff
@@ -0,0 +1,530 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "precomp.h"
+
+#if defined(EVENT_TRACING)
+#include "Device.tmh"
+#endif
+
+#ifdef ALLOC_PRAGMA
+#pragma alloc_text(PAGE, ViomemEvtDeviceContextCleanup)
+#pragma alloc_text(PAGE, ViomemEvtDevicePrepareHardware)
+#pragma alloc_text(PAGE, ViomemEvtDeviceReleaseHardware)
+#pragma alloc_text(PAGE, ViomemEvtDeviceD0Exit)
+#pragma alloc_text(PAGE, ViomemEvtDeviceD0ExitPreInterruptsDisabled)
+#pragma alloc_text(PAGE, ViomemDeviceAdd)
+#pragma alloc_text(PAGE, ViomemCloseWorkerThread)
+
+#endif // ALLOC_PRAGMA
+
+
+NTSTATUS
+ViomemDeviceAdd(
+    IN WDFDRIVER  Driver,
+    IN PWDFDEVICE_INIT  DeviceInit)
+{
+    NTSTATUS                     status = STATUS_SUCCESS;
+    WDFDEVICE                    device;
+    PDEVICE_CONTEXT              devCtx = NULL;
+    WDF_INTERRUPT_CONFIG         interruptConfig;
+    WDF_OBJECT_ATTRIBUTES        attributes;
+    WDF_PNPPOWER_EVENT_CALLBACKS pnpPowerCallbacks;
+
+    UNREFERENCED_PARAMETER(Driver);
+    PAGED_CODE();
+
+    TraceEvents(TRACE_LEVEL_INFORMATION, DBG_PNP, "%s Entry\n", __FUNCTION__);
+
+    WDF_PNPPOWER_EVENT_CALLBACKS_INIT(&pnpPowerCallbacks);
+
+    pnpPowerCallbacks.EvtDevicePrepareHardware      = ViomemEvtDevicePrepareHardware;
+    pnpPowerCallbacks.EvtDeviceReleaseHardware      = ViomemEvtDeviceReleaseHardware;
+    pnpPowerCallbacks.EvtDeviceD0Entry              = ViomemEvtDeviceD0Entry;
+    pnpPowerCallbacks.EvtDeviceD0Exit               = ViomemEvtDeviceD0Exit;
+    pnpPowerCallbacks.EvtDeviceD0ExitPreInterruptsDisabled = ViomemEvtDeviceD0ExitPreInterruptsDisabled;
+
+    WdfDeviceInitSetPnpPowerEventCallbacks(DeviceInit, &pnpPowerCallbacks);
+
+    WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&attributes, DEVICE_CONTEXT);
+    attributes.EvtCleanupCallback = ViomemEvtDeviceContextCleanup;
+
+    /* The driver initializes all the queues under lock of
+     * WDF object of the device. If we use default execution
+     * level, this lock is spinlock and common blocks required
+     * for queues can't be allocated on DISPATCH. So, we change
+     * the execution level to PASSIVE -> the lock is fast mutex
+     */
+    attributes.ExecutionLevel = WdfExecutionLevelPassive;
+
+    status = WdfDeviceCreate(&DeviceInit, &attributes, &device);
+    if(!NT_SUCCESS(status))
+    {
+        TraceEvents(TRACE_LEVEL_ERROR, DBG_PNP,
+           "WdfDeviceCreate failed with status 0x%08x\n", status);
+        return status;
+    }
+
+    devCtx = GetDeviceContext(device);
+
+    WDF_INTERRUPT_CONFIG_INIT(&interruptConfig,
+                            ViomemISR,
+                            ViomemDPC);
+
+    interruptConfig.EvtInterruptEnable  = ViomemEnableInterrupts;
+    interruptConfig.EvtInterruptDisable = ViomemDisableInterrupts;
+
+	// Is this usable??
+    //interruptConfig.AutomaticSerialization = TRUE;

```

Comment:
Probably useful for WHQL testing. On other hand, no userspace app that will use the driver, I am right?